src/checkers/inference/solver/backend/z3smt/Z3SmtSolver.java

```diff
@@ -0,0 +1,429 @@
+package checkers.inference.solver.backend.z3smt;
+
+import checkers.inference.model.ArithmeticConstraint;
+import checkers.inference.model.ArithmeticConstraint.ArithmeticOperationKind;
+import checkers.inference.model.Constraint;
+import checkers.inference.model.Slot;
+import checkers.inference.model.serialization.ToStringSerializer;
+import checkers.inference.solver.backend.Solver;
+import checkers.inference.solver.backend.z3smt.encoder.Z3SmtSoftConstraintEncoder;
+import checkers.inference.solver.frontend.Lattice;
+import checkers.inference.solver.util.ExternalSolverUtils;
+import checkers.inference.solver.util.FileUtils;
+import checkers.inference.solver.util.SolverArg;
+import checkers.inference.solver.util.SolverEnvironment;
+import checkers.inference.solver.util.Statistics;
+import com.microsoft.z3.BoolExpr;
+import com.microsoft.z3.Context;
+import com.microsoft.z3.Expr;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.logging.Logger;
+import javax.lang.model.element.AnnotationMirror;
+import org.checkerframework.javacutil.BugInCF;
+
+public class Z3SmtSolver<SlotEncodingT, SlotSolutionT>
+        extends Solver<Z3SmtFormatTranslator<SlotEncodingT, SlotSolutionT>> {
+
+    public enum Z3SolverEngineArg implements SolverArg {
+        /** option to use optimizing mode or not */
+        optimizingMode
+    }
+
+    private static final Logger logger = Logger.getLogger(Z3SmtSolver.class.getName());
+
+    protected final Context ctx;
+    protected com.microsoft.z3.Optimize solver;
+    protected StringBuilder smtFileContents;
+
+    protected static final String z3Program = "z3";
+    protected boolean optimizingMode;
+
+    /** This field indicates that whether we are going to explain unsatisfiable.*/
+    protected boolean getUnsatCore;
+
+    // used in non-optimizing mode to find unsat constraints
+    protected final Map<String, Constraint> serializedConstraints = new HashMap<>();
+
+    /** The set of IDs of unsat core constraints */
+    protected final List<String> unsatConstraintIDs = new ArrayList<>();
```

Comment:
This is also done rather confusingly. Instead of putting this in a field, make this the return of `runZ3Solver` in the unsat case. `parseStdOut` can take another argument for this list and then depending on sat/unsat either `result` or `unsatContraintIDs` is returned by `runZ3Solver`. Oh, but then we need some other way to decide whether it's sat or unsat. Please have a look whether you can make this easier to follow.

Reply:
How about `runZ3Solver` returns boolean indicating sat/unsat, and takes the `List<String>` as output parameter?  (`List` is mutable)